src/file-engine/src/manifest.rs

```diff
@@ -0,0 +1,129 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use common_datasource::file_format::Format;
+use object_store::ObjectStore;
+use serde::{Deserialize, Serialize};
+use snafu::{ensure, OptionExt, ResultExt};
+use store_api::metadata::{ColumnMetadata, RegionMetadataBuilder, RegionMetadataRef};
+use store_api::storage::{ColumnId, RegionId};
+
+use crate::error::{
+    CheckObjectSnafu, DecodeJsonSnafu, DeleteRegionManifestSnafu, EncodeJsonSnafu,
+    InvalidMetadataSnafu, LoadRegionManifestSnafu, ManifestExistsSnafu, MissingRequiredFieldSnafu,
+    ParseFileFormatSnafu, Result, StoreRegionManifestSnafu,
+};
+use crate::FileOptions;
+
+#[inline]
+fn region_manifest_path(region_dir: &str) -> String {
+    format!("{region_dir}manifest/_immutable_manifest")
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct FileRegionManifest {
+    pub region_id: RegionId,
+    pub column_metadatas: Vec<ColumnMetadata>,
+    pub primary_key: Vec<ColumnId>,
+    pub options: HashMap<String, String>,
+}
+
+impl FileRegionManifest {
+    pub async fn store(&self, region_dir: &str, object_store: &ObjectStore) -> Result<()> {
+        let path = &region_manifest_path(region_dir);
+        let exist = object_store
+            .is_exist(path)
+            .await
+            .context(CheckObjectSnafu { path })?;
+        ensure!(!exist, ManifestExistsSnafu { path });
+
+        let bs = self.encode()?;
+        object_store
+            .write(path, bs)
+            .await
+            .context(StoreRegionManifestSnafu {
+                region_id: self.region_id,
+            })?;
+
+        Ok(())
+    }
+
+    pub async fn load(
+        region_id: RegionId,
+        region_dir: &str,
+        object_store: &ObjectStore,
+    ) -> Result<Self> {
+        let path = &region_manifest_path(region_dir);
+        let bs = object_store
+            .read(path)
+            .await
+            .context(LoadRegionManifestSnafu { region_id })?;
+        Self::decode(bs.as_slice())
+    }
+
+    pub async fn delete(
+        region_id: RegionId,
+        region_dir: &str,
+        object_store: &ObjectStore,
+    ) -> Result<()> {
+        let path = &region_manifest_path(region_dir);
+        object_store
+            .delete(path)
+            .await
+            .context(DeleteRegionManifestSnafu { region_id })
+    }
+
+    pub fn metadata(&self) -> Result<RegionMetadataRef> {
+        let mut builder = RegionMetadataBuilder::new(self.region_id);
+        for column in &self.column_metadatas {
+            builder.push_column_metadata(column.clone());
+        }
+        builder.primary_key(self.primary_key.clone());
+        let metadata = builder.build().context(InvalidMetadataSnafu)?;
```

Comment:
The builder will fail if we don't contain any columns that have the `Timestamp` semantic.